### ISSUE TITLE
RFE: Two-way audio

### ISSUE THREAD

**Issue by obriencj** [briis/unifiprotect#148]
Thank you so much for this integration!

I just wanted to add a tracker feature request to enable two-way audio to a supported protect device, such as a G4 Doorbell.

**Comment by briis** [comment]
Good suggestion. I am not sure how to start on this, but now it is added as a feature request, and I will look in to this. Or if someone else wants to grab this, feel free.

**Comment by AngellusMortis** [comment]
I would love to see this as well, but I am pretty sure the feature is locked down to the Unifi Protect app.

**Comment by obriencj** [comment]
The homebridge plugin claims to have implemented it, but I have not personally verified the functionality.
https://github.com/hjdhjd/homebridge-unifi-protect/blob/master/docs/Doorbell.md#doorbell-support

It seems to be implemented using RTP (or RTC or RTCP? only familiarity was in looking through their sources briefly)

**Comment by briis** [comment]
I know the Homebridge implementation has this implemented, so it might be possible to do.But currently I have no clue how how this can be implemented.

But let us play with idea - What is the use scenario in HA for this?

**Comment by AngellusMortis** [comment]
It is really simple: the Unifi Protect app requires cloud access to be enabled to use. 

Also, centralizing access to features in HA instead of Unifi Protect means it is easier to manage accounts and access for other people in my household.

**Comment by briis** [comment]
> the Unifi Protect app requires cloud access to be enabled to use.

This is to my knowledge not true. The latest version of the app will try direct access first, and only use cloud if it cannot connect using the LAN.

> centralizing access to features in HA instead of Unifi Protect means it is easier to manage accounts and access for other people in my household.

This one makes sense.

How would you interface with two-way audio through Home Assistant? I have not see a Lovelace card or similar that supports that?

**Comment by obriencj** [comment]
A lovelace card would be my first guess there, yes. Similarly to how we can see a video stream today. Maybe implemented via webrtc?

**Comment by AngellusMortis** [comment]
> This is to my knowledge not true. The latest version of the app will try direct access first, and only use cloud if it cannot connect using the LAN.

I have been fighting with the app for the past 2 days. As soon as I disable Remote Access on my UDM-Pro, the Unifi Protect app stops working. It cannot see my controller. It seems like if I enable Remote Access, it will use my local connection to the camera instead of an over the Internet one. I am on the latest version of the Android app as well (1.3.11). 

> How would you interface with two-way audio through Home Assistant?

And yeah, probably via a Lovelace card. It would probably be a lot of custom frontend code to implement, but I would imagine just a button that turns on your microphone and sends the audio to the speaker. If comparing it to the picture card with the Camera live feed modal, it would probably only be in the modal popup.

**Comment by briis** [comment]
For the IOS App, which I use, this works. I have not disabled remote connection, I can just see that when I am at home, it connects directly via the Internal IP, and uses the Cloud when I am outside my LAN.

**Comment by briis** [comment]
Anyway, I would really love to have this feature in the Integration, but don't expect it right away.
I can definitely get some good ideas from the HomeBridge version, but there is still a lot I need to figure out with regards to sending the audio forth and back to Home Assistant.

**Comment by obriencj** [comment]
Thank you!

**Comment by Flight777** [comment]
I'm really +1 this request as well :)

Use case:

We have a Home Assistant dashboard (tablet) on the wall in the livingroom. When someone rings the doorbell, it pop-ups with the camera feed and some buttons for the text, would be really nice if you can also just hold a mic button to directly talk or something.

Maybe as an in between step: would be already really cool if you can send (instead of text only) some pre-recorded or TTS message to the doorbell that is then plays over the speaker. Because the text is really hard to notice for persons, so if you have some pre recorded audio asking to put the package somewhere, that already helps a ton 👍

**Comment by briis** [comment]
> Maybe as an in between step: would be already really cool if you can send (instead of text only) some pre-recorded or TTS 

Yes, that could be a solution, but in order to do this, the doorbell needs to be exposed as a media_player, and this is currently the issue for me - how to do that.

**Comment by hjdhjd** [comment]
As the developer of the Homebridge plugin...happy to explain how it works. In essence, you send an AAC stream directly to the doorbell via ADTS. Here’s the key ffmpeg magic incantation you need to make it all happen:

https://github.com/hjdhjd/homebridge-unifi-protect/blob/f236df00a4a47ceb8ae5ca3a84305ff786fc4f23/src/protect-stream.ts#L487

One key thing to note, and it’s essential to understand, unlike virtually everything else in Protect, this goes directly to the doorbell, **not** the controller.

There’s one notable caveat, and that relates to AEC. See https://github.com/hjdhjd/homebridge-unifi-protect/blob/master/docs/Doorbell.md - particularly the section detailing two-way audio considerations.

Best of luck!

**Comment by briis** [comment]
Thank you @hjdhjd Really appreciate your help. I have looked at what you did already and I will need to figure out how to implement this in HA. Good tip on the AAC stream going directly to the doorbell.

**Comment by adrum** [comment]
It would be really cool to find a way to do this not only from a lovelace card, but also from the HomeKit integration in HA. I wonder if there is a way to this that make it compatible with both.

**Comment by georgeboot** [comment]
> but also from the HomeKit integration in HA

I actually installed this integration today and saw the "speak" button in homekit when viewing the stream. Turns out, it does nothing, but HomeKit seems to believe the doorbel has the functionality.

**Comment by Jpeterson37** [comment]
Just waned to chime in and say if this could be added as a media player also. If could support TTS for automated responses and media playback for Halloween

**Comment by AngellusMortis** [comment]
This issue is _just_ for live way talk. There is separate issue for a media player / TTS support (https://github.com/briis/unifiprotect/issues/304). It is already in progress as well. We plan to add it to 0.11
